lib/metrician/reporters/active_record.rb

```diff
@@ -23,7 +23,24 @@ module QueryInterceptor
     QUERY_METRIC = "database.query"
 
     def self.included(instrumented_class)
-      return if instrumented_class.method_defined?(:log_without_metrician)
+      # You might wonder why this exception-based flow control is here instead of
+      # something more sane like
+      # instrumented_class.method_defined?(:log_without_metrician)
+      # The answer is that method_defined?, methods, instance_methods
+      # and other reasonable ways of answering this question do not
+      # work correctly in this circumstance.  They return false even
+      # when log_without_metrician is defined on the class.  The
+      # result is that the guard statement fails and runs this code
+      # twice.  That makes log_without_metrician an alias of
+      # log_with_metrician and causes infinite recursion :(
+      already_defined = true
+      begin
+        already_defined = instrumented_class.new(nil).method(:log_without_metrician)
+      rescue
+        already_defined = false
+      end
+
+      return if already_defined
```

Comment:
It looks like `private_method_defined?` is the way to get at this particular method here, if we don't want to use exception based flow control:
```
   39:       rescue
   40:         already_defined = false
   41:       end
   42:       debugger if !instrumented_class.method_defined?(:log_without_metrician) && already_defined
   43:
=> 44:       return if already_defined
   45:       debugger
   46:       instrumented_class.class_eval do
   47:         debugger
   48:         alias_method :log_without_metrician, :log
(byebug) instrumented_class.method_defined?(:log_without_metrician)
false
(byebug) instrumented_class.private_method_defined?(:log_without_metrician)
true
```

Maybe the way that the `log` method gets marked as private has changed in Rails and our alias is ending up marked as private? I could see checking both `method_defined?` and `private_method_defined?` to be sure.

I'm pro switching this, but I'm happy to talk about it more.

Reply:
`private_method_defined?` does appear to work for some of the ruby/rails combos, but not all of them. I'm just doing a quick check, so it doesn't work for the oldest combo and it does work on the newest one. 

Pros and cons

1) Attempting to get the unbound method definitely works across all versions. It will probably work forever because if you can't access a method through `method` then there is no method. But, you have to use exception flow control to use `method`.

2) We can try to figure out which methods work with which combos and conditionalize it. I think it's probably a pre-2.5/post-2.5 split. 

3) We can try to verify that some combination of interrogation methods should cover every combo (e.g. `method_defined?` and `private_method_defined?`)

Given that the outcome is infinite recursion I prefer the one that definitely can't be wrong, but I'm convincible. It's not like I think exception-based flow control is the bees knees.

Reply:
I don't feel super strongly either way. The fact that `*method_defined?` has changed over time does make me wonder if `method` will actually be trustworthy into the future, but probably. If we leave the exception based flow control we might want to move it into and `already_defined?` method just to encapsulate the badness.

In any case I think we should definitely add some kind of a test for this.